File: Cryambly/Engine/Models/Characters/Attachments/AttachmentManager.cs
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using CryCil.Engine.Physics;
using CryCil.Hashing;

namespace CryCil.Engine.Models.Characters.Attachments
{
	/// <summary>
	/// Represents an object that manages objects that are attached to the character.
	/// </summary>
	[StructLayout(LayoutKind.Explicit)]
	public struct AttachmentManager
	{
		#region Fields
		[FieldOffset(0)] private readonly IntPtr handle;
		/// <summary>
		/// Provides access to the collection of attachments.
		/// </summary>
		[FieldOffset(0)] public readonly AttachmentManagerSockets Sockets;
		/// <summary>
		/// Provides access to the collection of attachment proxies.
		/// </summary>
		[FieldOffset(0)] public readonly AttachmentManagerProxies Proxies;
		#endregion
		#region Properties
		/// <summary>
		/// Indicates whether this instance is usable.
		/// </summary>
		public bool IsValid
		{
			get { return this.handle != IntPtr.Zero; }
		}
		/// <summary>
		/// Gets the character this object manages attachments and attachment proxies for.
		/// </summary>
		/// <exception cref="NullReferenceException">This instance is not valid.</exception>
		public Character Character
		{
			get
			{
				this.AssertInstance();

				return GetSkelInstance(this.handle);
			}
		}
		#endregion
		#region Construction
		internal AttachmentManager(IntPtr handle)
			: this()
		{
			this.handle = handle;
		}
		#endregion
		#region Utilities
		/// <exception cref="NullReferenceException">This instance is not valid.</exception>
		private void AssertInstance()
		{
			if (!this.IsValid)
			{
				throw new NullReferenceException("This instance is not valid.");
			}
		}

		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern uint LoadAttachmentList(IntPtr handle, string pathname);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern AttachmentSocket CreateAttachment(IntPtr handle, string szName, AttachmentTypes type,
																 string szJointName, bool bCallProject);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern int RemoveAttachmentByInterface(IntPtr handle, AttachmentSocket ptr);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern int RemoveAttachmentByName(IntPtr handle, string szName);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern int RemoveAttachmentByNameCrc(IntPtr handle, LowerCaseCrc32 nameCrc);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern AttachmentSocket GetInterfaceByName(IntPtr handle, string szName);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern AttachmentSocket GetInterfaceByIndex(IntPtr handle, uint c);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern AttachmentSocket GetInterfaceByNameCrc(IntPtr handle, LowerCaseCrc32 nameCrc);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern int GetAttachmentCount(IntPtr handle);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern int GetIndexByName(IntPtr handle, string szName);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern int GetIndexByNameCrc(IntPtr handle, LowerCaseCrc32 nameCrc);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern uint ProjectAllAttachment(IntPtr handle);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern void PhysicalizeAttachment(IntPtr handle, int idx, PhysicalEntity pent, int nLod);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern void DephysicalizeAttachment(IntPtr handle, int idx, PhysicalEntity pent);
		[MethodImpl(MethodImplOptions.InternalCall)]
		private static extern Character GetSkelInstance(IntPtr handle);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern int GetProxyCount(IntPtr handle);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern AttachmentProxy CreateProxy(IntPtr handle, string szName, string szJointName);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern AttachmentProxy GetProxyInterfaceByIndex(IntPtr handle, uint c);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern AttachmentProxy GetProxyInterfaceByName(IntPtr handle, string szName);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern AttachmentProxy GetProxyInterfaceByCrc(IntPtr handle, LowerCaseCrc32 nameCrc);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern int GetProxyIndexByName(IntPtr handle, string szName);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern int RemoveProxyByInterface(IntPtr handle, AttachmentProxy ptr);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern int RemoveProxyByName(IntPtr handle, string szName);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern int RemoveProxyByNameCrc(IntPtr handle, LowerCaseCrc32 nameCrc);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern void DrawProxies(IntPtr handle, uint enable);
		#endregion
	}
}

File: MonoInterface/Interops/AttachmentManager.cpp
#include "stdafx.h"

#include "AttachmentManager.h"
#include <IAttachment.h>

void AttachmentManagerInterop::OnRunTimeInitialized()
{
	REGISTER_METHOD(GetSkelInstance);
	REGISTER_METHOD(LoadAttachmentList);
	REGISTER_METHOD(CreateAttachment);
	REGISTER_METHOD(RemoveAttachmentByInterface);
	REGISTER_METHOD(RemoveAttachmentByName);
	REGISTER_METHOD(RemoveAttachmentByNameCrc);
	REGISTER_METHOD(GetInterfaceByName);
	REGISTER_METHOD(GetInterfaceByIndex);
	REGISTER_METHOD(GetInterfaceByNameCrc);
	REGISTER_METHOD(GetAttachmentCount);
	REGISTER_METHOD(GetIndexByName);
	REGISTER_METHOD(GetIndexByNameCrc);
	REGISTER_METHOD(ProjectAllAttachment);
	REGISTER_METHOD(PhysicalizeAttachment);
	REGISTER_METHOD(DephysicalizeAttachment);
	REGISTER_METHOD(GetProxyCount);
	REGISTER_METHOD(CreateProxy);
	REGISTER_METHOD(GetProxyInterfaceByIndex);
	REGISTER_METHOD(GetProxyInterfaceByName);
	REGISTER_METHOD(GetProxyInterfaceByCrc);
	REGISTER_METHOD(GetProxyIndexByName);
	REGISTER_METHOD(RemoveProxyByInterface);
	REGISTER_METHOD(RemoveProxyByName);
	REGISTER_METHOD(RemoveProxyByNameCrc);
	REGISTER_METHOD(DrawProxies);
}

ICharacterInstance *AttachmentManagerInterop::GetSkelInstance(IAttachmentManager *handle)
{
	return handle->GetSkelInstance();
}

uint AttachmentManagerInterop::LoadAttachmentList(IAttachmentManager *handle, mono::string pathname)
{
	return handle->LoadAttachmentList(NtText(pathname));
}

IAttachment *AttachmentManagerInterop::CreateAttachment(IAttachmentManager *handle, mono::string szName, uint type, mono::string szJointName, bool bCallProject)
{
	return handle->CreateAttachment(NtText(szName), type, NtText(szJointName), bCallProject);
}

int AttachmentManagerInterop::RemoveAttachmentByInterface(IAttachmentManager *handle, IAttachment *ptr)
{
	return handle->RemoveAttachmentByInterface(ptr);
}

int AttachmentManagerInterop::RemoveAttachmentByName(IAttachmentManager *handle, mono::string szName)
{
	return handle->RemoveAttachmentByName(NtText(szName));
}

int AttachmentManagerInterop::RemoveAttachmentByNameCrc(IAttachmentManager *handle, uint nameCrc)
{
	return handle->RemoveAttachmentByNameCRC(nameCrc);
}

IAttachment *AttachmentManagerInterop::GetInterfaceByName(IAttachmentManager *handle, mono::string szName)
{
	return handle->GetInterfaceByName(NtText(szName));
}

IAttachment *AttachmentManagerInterop::GetInterfaceByIndex(IAttachmentManager *handle, uint c)
{
	return handle->GetInterfaceByIndex(c);
}

IAttachment *AttachmentManagerInterop::GetInterfaceByNameCrc(IAttachmentManager *handle, uint nameCrc)
{
	return handle->GetInterfaceByNameCRC(nameCrc);
}

int AttachmentManagerInterop::GetAttachmentCount(IAttachmentManager *handle)
{
	return handle->GetAttachmentCount();
}

int AttachmentManagerInterop::GetIndexByName(IAttachmentManager *handle, mono::string szName)
{
	return handle->GetIndexByName(NtText(szName));
}

int AttachmentManagerInterop::GetIndexByNameCrc(IAttachmentManager *handle, uint nameCrc)
{
	return handle->GetIndexByNameCRC(nameCrc);
}

uint AttachmentManagerInterop::ProjectAllAttachment(IAttachmentManager *handle)
{
	return handle->ProjectAllAttachment();
}

void AttachmentManagerInterop::PhysicalizeAttachment(IAttachmentManager *handle, int idx, IPhysicalEntity *pent, int nLod)
{
	handle->PhysicalizeAttachment(idx, pent, nLod);
}

void AttachmentManagerInterop::DephysicalizeAttachment(IAttachmentManager *handle, int idx, IPhysicalEntity *pent)
{
	handle->DephysicalizeAttachment(idx, pent);
}

int AttachmentManagerInterop::GetProxyCount(IAttachmentManager *handle)
{
	return handle->GetProxyCount();
}

IProxy *AttachmentManagerInterop::CreateProxy(IAttachmentManager *handle, mono::string szName, mono::string szJointName)
{
	return handle->CreateProxy(NtText(szName), NtText(szJointName));
}

IProxy *AttachmentManagerInterop::GetProxyInterfaceByIndex(IAttachmentManager *handle, uint c)
{
	return handle->GetProxyInterfaceByIndex(c);
}

IProxy *AttachmentManagerInterop::GetProxyInterfaceByName(IAttachmentManager *handle, mono::string szName)
{
	return handle->GetProxyInterfaceByName(NtText(szName));
}

IProxy *AttachmentManagerInterop::GetProxyInterfaceByCrc(IAttachmentManager *handle, uint nameCrc)
{
	return handle->GetProxyInterfaceByCRC(nameCrc);
}

int AttachmentManagerInterop::GetProxyIndexByName(IAttachmentManager *handle, mono::string szName)
{
	return handle->GetProxyIndexByName(NtText(szName));
}

int AttachmentManagerInterop::RemoveProxyByInterface(IAttachmentManager *handle, IProxy *ptr)
{
	return handle->RemoveProxyByInterface(ptr);
}

int AttachmentManagerInterop::RemoveProxyByName(IAttachmentManager *handle, mono::string szName)
{
	return handle->RemoveProxyByName(NtText(szName));
}

int AttachmentManagerInterop::RemoveProxyByNameCrc(IAttachmentManager *handle, uint nameCrc)
{
	return handle->RemoveProxyByNameCRC(nameCrc);
}

void AttachmentManagerInterop::DrawProxies(IAttachmentManager *handle, uint enable)
{
	handle->DrawProxies(enable);
}
